[app][feat] use config file for logging (#460)
OCR-D/core: ocrd_utils/README.md

# ocrd_utils

> Additional functionalities for OCR-D framework

Reviewer: ```suggestion
> OCR-D framework - shared code, helpers, constants
```

Besides, this quotation is identical to the setuptools description.

See https://github.com/OCR-D/core

## OCR-D Module Logging

> _Please note, that the following instructions only apply to the dockerized version of `ocrd_utils`._
> 
> Configuration of Logging-Facilities is done with standard Python3 Logging Library from modul `logging`. This way, logging configuration can be customized for all OCR-D modules using the Python logging framework.  
> A sample configuration file is included in the OCR-D container image at `/etc/ocrd_logging.conf`. This cofiguration is also meant to be the default configuration for OCR-D container images.  
> If this file doesn't exist or gets removed, logging mechanics use as fallback the `logging.basicConfig` functionalities which use stdout as message channel. 
> 
> For more information about logging, handlers and formats, please see https://docs.python.org/3/howto/logging.htm.
> 
> ### Customization
> 
> Create a custom Configuration and include this as part of your local workspace when running the OCR-D container.  
> In this case, it will override *all* settings from default configuration, which is included at `/etc/ocrd_logging.conf`.  

Reviewer: I don't think we should make this about Docker containers specifically. First we need a documentation of the general mechanism here (as in my suggestion for the preamble of the template file, but more user-oriented). After that, some specifics for Docker setup can be given.

```suggestion
File-based control over logging facilities is done with standard [Python 3 logging module configuration files](https://docs.python.org/3.6/howto/logging.html#configuring-logging). This way, the level, format and destinations of log messages can be customized for all OCR-D modules individually and persistently, in the usual syntax.

A template configuration file (with commented examples) is included in [ocrd_logging.conf](./ocrd_logging.conf). This is meant as an example, and should be **customized**. 

To get into effect, you must put a copy (under the same name) into:
1. your current working directory, 
2. your user directory, or
3. `/etc`. 
These directories are searched in said order, and the first find wins. When no config file is found, the default logging configuration applies (which uses only stdout and the `INFO` loglevel for most loggers, cf. [here](./ocrd_logging.py)).

Thus, a configuration file will override *all* settings from the default configuration, and from configuration files in lower-priority directories.

For more information about logging, handlers and formats, see [Python documentation](https://docs.python.org/3/howto/logging.htm).

#### Docker containers

In the Dockerfiles used to build `ocrd/core` (and subsequently `ocrd/all`), the above mentioned template is directly copied to `/etc/ocrd_logging.conf` within the container image. This cofiguration is thereby also the default configuration for OCR-D containers. 

Thus, if you want to customize logging rules in one of these Docker containers, you can create a custom configuration file and either:
- place it into your local workspace directory when running the OCR-D container.
- include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.
```

Author: Well, a really smart suggestion!

To get into effect, you must put a copy (under the same name) into:
1. your current working directory, 
2. your user directory, or
3. `/etc`. 
These directories are searched in said order, and the first find wins. When no config file is found, the default logging configuration applies (which uses only stdout and the `INFO` loglevel for most loggers, cf. [here](./ocrd_logging.py)).

Thus, a configuration file will override *all* settings from the default configuration, and from configuration files in lower-priority directories.

For more information about logging, handlers and formats, see [Python documentation](https://docs.python.org/3/howto/logging.htm).

#### Docker containers

In the Dockerfiles used to build `ocrd/core` (and subsequently `ocrd/all`), the above mentioned template is directly copied to `/etc/ocrd_logging.conf` within the container image. This cofiguration is thereby also the default configuration for OCR-D containers. 

Thus, if you want to customize logging rules in one of these Docker containers, you can create a custom configuration file and either:

> - place it into your local workspace directory when running the OCR-D container.

Reviewer: ```suggestion
- place it into your local workspace directory when running the OCR-D container.
- mount it under `/etc` when starting up the container, e.g. `docker run --mount type=bind,source=host/path/to/your-template.conf,destination=/etc/ocrd_logging.conf ocrd/all`
```

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Reviewer: > Couldn't you mount the file into a container at runtime as well?

Theoretically yes. But unlikely with our search directories (except if the mount point is the workspace directory itself, which is already covered by the description): `$HOME` will almost always resolve to `/` because we must keep the external user for FS access, and `/etc` would probably break most tools running in the container.

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Reviewer 2: You can mount individual files into a docker container, at least it was possible in the past. But we could also extend the search path to include `/etc/ocrd/ocrd_logging.conf` so a directory with the logging configuration file (and possibly other configuration files in the future) could be mounted.

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Author: I guess the point is, when you copy it into a running container, it can cause strange issues since for the first modules the default config will be used immediately, but further ones might use the new copy. So we might end up with two loggers.

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Reviewer: > when you copy it into a running container, it can cause strange issues since for the first modules the default config will be used immediately, but further ones might use the new copy. So we might end up with two loggers.

What do you mean by _running container_ – a running workflow, or just that the Docker has been spawned from the image? The latter case would not be any different from copying the file at build time, and the former case is just bad luck/management (and still would _not_ yield two loggers at the same time)...

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Reviewer 2: I did not mean "running container", that would be strange indeed. I mean at the point when `docker run` is executed, you can define volumes/binds. All the files have to be mounted before the `ENTRYPOINT` is executed.

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Author: A running workflow (inside a container), and yes, I agree, this is an indication of organizational failure ;)

> - include a Dockerfile step (layer or stage) which copies this into `/etc/ocrd_logging.conf` at build time in your own Docker image.

Reviewer: > I mean at the point when `docker run` is executed, you can define volumes/binds. All the files have to be mounted before the `ENTRYPOINT` is executed.

I know. (That's why thumbs are up.) So IMO this realistic option should be added as a second item in the description.
